File: java/L900/LeetCode0987.java
package L900;

import java.util.ArrayList;
import java.util.Collections;
import java.util.Deque;
import java.util.HashMap;
import java.util.LinkedList;
import java.util.List;
import java.util.Map;
import java.util.stream.Collectors;

import util.TreeNode;

public class LeetCode0987 {
    private class Node {
        TreeNode tNode;
        int col;

        public Node(TreeNode node, int c) {
            this.tNode = node;
            this.col = c;
        }
    }

    public List<List<Integer>> verticalTraversal(TreeNode root) {
        if (root == null) {
            return new ArrayList<>();
        }
        HashMap<Integer, List<Integer>> rowMap = new HashMap<>();
        Deque<Node> queue = new LinkedList<>();
        queue.offer(new Node(root, 0));
        while (!queue.isEmpty()) {
            HashMap<Integer, List<Integer>> rows = new HashMap<>();
            int sz = queue.size();
            for (int i = 0;i < sz;++i) {
                Node node = queue.poll();
                List<Integer> cList = rows.getOrDefault(node.col, new ArrayList<>());
                cList.add(node.tNode.val);
                rows.put(node.col, cList);
                if (node.tNode.left != null) {
                    queue.offer(new Node(node.tNode.left, node.col - 1));
                }
                if (node.tNode.right != null) {
                    queue.offer(new Node(node.tNode.right, node.col + 1));
                }
            }
            for (Map.Entry<Integer, List<Integer>> entry : rows.entrySet()) {
                List<Integer> cList = rowMap.getOrDefault(entry.getKey(), new ArrayList<>());
                Collections.sort(entry.getValue());
                cList.addAll(entry.getValue());
                rowMap.put(entry.getKey(), cList);
            }
        }
        return rowMap.entrySet().stream().sorted((a, b) -> {
            return a.getKey() - b.getKey();
        }).map(Map.Entry<Integer, List<Integer>>::getValue).collect(Collectors.toList());
    }
}

File: cpp/L900/LeetCode0987.cc
#include "../heads/tree_node.h"
#include <algorithm>
#include <map>
#include <utility>
#include <vector>
using std::vector;

class Solution {
    void traceTree(TreeNode *root, int row, int col, std::map<int, vector<std::pair<int, int>>> &mp) {
        if (root == nullptr) {
            return;
        }
        mp[col].emplace_back(std::pair<int, int>(row, root->val));
        traceTree(root->left, row + 1, col - 1, mp);
        traceTree(root->right, row + 1, col + 1, mp);
    }
public:
    vector<vector<int>> verticalTraversal(TreeNode* root) {
        std::map<int, vector<std::pair<int, int>>> mp;
        traceTree(root, 0, 0, mp);
        vector<vector<int>> res;
        for (auto it{mp.begin()};it != mp.end();++it) {
            std::sort(it->second.begin(), it->second.end(), [](std::pair<int, int> const &p1, std::pair<int, int> const &p2) {
                return p1.first == p2.first ? p1.second < p2.second : p1.first < p2.first;
            });
            vector<int> r(it->second.size());
            std::transform(it->second.begin(), it->second.end(), r.begin(), [](std::pair<int, int> const &p) {
                return p.second;
            });
            res.push_back(r);
        }
        return res;
    }
};